set.mm

```diff
@@ -730651,6 +730957,1157 @@ last digit of the integer part (for ` b = ` 10 the first digit before
+      nnncan2d eqeq2d 3bitr2d ) ACUABUBZPQZDRQVMERQSQUCFVNGRQVMHRQSQUCUDBUEUFFH
+      GPQZEDPQZUGQZCDPQZRQZGSQZUCVPFRQZVPVTRQZUCWAVOCRQZGERQZDHRQZPQZSQZUCABCDE
+      VQFGHIJKLMNOVQUTUHAFVTVPAFMUJAVSGAVQVRAVOVPAHGAHOUJZAGNUJZUIZAEDAEKUJZADJ
+      UJZUIZAEDWKWLADELUKULZUMZACDACIUJZWLUIZTZWIUNWMWNUOAWBWGWAAWBVPVSRQZVPGRQ
+      ZSQZWCEGRQZDGRQZPQZVODRQZPQZSQZWGAVPVSGWMWRWIUPAXAWCXEPQZXDSQXGAWSXHWTXDS
+      AVPVQRQZVRRQVOVRRQWSXHAXIVOVRRAVOVPWJWMWNUQURAVPVQVRWMWOWQUSAVOCDWJWPWLVA
+      VBAEDGWKWLWIVCVDAWCXEXDAVOCWJWPTAVODWJWLTAXBXCAEGWKWITADGWLWITUIVEVFAXFWF
+      WCSAXFWDGDRQZPQZHDRQZXJPQZPQWDXLPQWFAXDXKXEXMPAXBWDXCXJPAEGWKWIVGADGWLWIV
+      GVDAHGDWHWIWLVCVDAWDXLXJAGEWIWKTAHDWHWLTAGDWIWLTVJAXLWEWDPAHDWHWLVGVHVIVH
+      VIVKVL $.
+  $}
+
+  ${
+    affineid.f $e |- ( ph -> A e. CC ) $.
+    affineid.x $e |- ( ph -> T e. CC ) $.
+    $( Identity of an affine combination.  (Contributed by AV, 2-Feb-2023.) $)
+    affineid $p |- ( ph -> ( ( ( 1 - T ) x. A ) + ( T x. A ) ) = A ) $=
+      ( c1 cmin co cmul caddc 1cnd subdird mulid2d oveq1d eqtrd mulcld npcand )
+      AFCGHBIHZCBIHZJHBSGHZSJHBARTSJARFBIHZSGHTAFCBAKEDLAUABSGABDMNONABSDACBEDP
+      QO $.
+  $}
+
+  ${
+    resum2sqcl.q $e |- Q = ( ( A ^ 2 ) + ( B ^ 2 ) ) $.
+    $( The sum of two squares of real numbers is a real number.  (Contributed
+       by AV, 7-Feb-2023.) $)
+    resum2sqcl $p |- ( ( A e. RR /\ B e. RR ) -> Q e. RR ) $=
+      ( cr wcel wa c2 cexp co caddc simpl resqcld simpr readdcld syl5eqel ) AEF
+      ZBEFZGZCAHIJZBHIJZKJEDSTUASAQRLMSBQRNMOP $.
+
+    $( The sum of the square of a nonzero real number and the square of another
+       real number is greater than zero.  (Contributed by AV, 7-Feb-2023.) $)
+    resum2sqgt0 $p |- ( ( ( A e. RR /\ A =/= 0 ) /\ B e. RR ) -> 0 < Q ) $=
+      ( cr wcel cc0 wne wa c2 co caddc clt simpl resqcld adantr simpr wbr sqgt0
+      cexp cle sqge0 adantl addgtge0d syl6breqr ) AEFZAGHZIZBEFZIZGAJTKZBJTKZLK
+      CMUJUKULUHUKEFUIUHAUFUGNOPUJBUHUIQOUHGUKMRUIASPUIGULUARUHBUBUCUDDUE $.
+  $}
+
+$(
+-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-
+  Spheres and lines in real Euclidean spaces
+-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-
+$)
+
+  $c LineM Sphere $.
+
+  $( Declare the syntax for lines in generalized real Euclidean spaces. $)
+  cline $a class LineM $.
+
+  $( Declare the syntax for spheres in generalized real Euclidean spaces. $)
+  csph $a class Sphere $.
+
+  ${
+    $d p t w x y $.
+    $( Definition of lines passing through two different points in a left
+       module (or any extended structure having a base set, an addition, and a
+       scalar multiplication).  (Contributed by AV, 14-Jan-2023.) $)
+    df-line $a |- LineM = ( w e. _V |-> ( x e. ( Base ` w ) ,
+                                         y e. ( ( Base ` w ) \ { x } )
+                    |-> { p e. ( Base ` w ) | E. t e. ( Base ` ( Scalar ` w ) )
```

Comment:
It does not seem good to have the set for the second argument depending on the value of the first. I would use
```
x e. Base w , y e. Base w |-> if ( x = y , ... , ... )
```
where the ... is some default value, probably `Base w` so that the value of the function is always: the set of z such that x,y,z are collinear.

Reply:
For reference, I could see this used at least in [`df-subma`](https://us.metamath.org/mpeuni/df-subma.html) and [`df-nbgr`](https://us.metamath.org/mpeuni/df-nbgr.html) (both by @avekens!)

I assume you would put the empty set `(/)` in the "else" part of the "if". That would make the definition specific to our encoding of "out-of-bounds behaviour". The proposed solution does not have that drawback and is shorter, IMO.

Reply:
Oops, I missed the second part of your comment!

Reply:
I think it was @digama0 who once proposed the style having a difference with a singleton. Furthermore, there are more examples of this kind (e.g. df-trkg by @tirix ;-) ). Finally, we have ~cbvmpt2x which allows the second class to be dependent on the first set variable.

Using this style, the if statement can be avoided, giving the same result (if x = y, then  ` ( x ( LineM ``  w ) y = (/)`in both cases.

Reply:
As I wrote above, choose a definition which is uniform in all cases, including when x=y. I gave one possiblity above: the set of z's such that x,y,z are collinear.

Actually, it would be better to define the "affine hull" of a family of points in all generality. Also, you defined it in terms of a particular parameterization (that is, the set of `(1-t)X+tY` for real t), which should be a property, not a definition, because it is not symmetric in x and y and has a different form for more points. The affine hull of a subset X of an R-affine space A is the set of (finite) linear combinations of elements of X such that the coefficients sum to $1_R$.

I have only one theorem on barycentric coordinates https://us.metamath.org/mpeuni/bj-bary1.html in the section "Affine geometry" but I think it gives the idea.

Then, if you consider spheres (so you leave affine geometry) in Euclidean spaces, you may generalize your results to the intersection of a max-dimensional sphere and an affine subspace. This is roughly the result that the metric induced on an affine subspace by a Euclidean metric is Euclidean.

Reply:
Because of the accordance with the definition of LineG I would leave it as it is.

Reply:
Your mathbox, your choice. Out of curiosity: where is the definition of lineG "df-lineg" gives nothing.

Reply:
The definition of LineG (as a slot index) is in df-lng.

Reply:
df-lng is not very enlightening...
It might be https://us.metamath.org/mpeuni/ttgval.html ? So for x=y it corresponds to the first proposal I made above (though now I think my second proposal using the affine hull generalizes more naturally to multiple points).

Reply:
@benjub you are right, the relationship between LineM and LineG is not very clear. Therefore, I added two references to important (defining) theorems in the comment of ~eenglngeehlnm.